src/test/scala/nodes/learning/PCASuite.scala

```diff
@@ -79,4 +81,49 @@ class PCATransformerSuite extends FunSuite with LocalSparkContext with Logging {
       assert(Stats.aboutEq(redCov(x,y), 0.0, 1e-6), s"PCA Matrix should be 0 off-diagonal. $x,$y = ${redCov(x,y)}")
     }
   }
+
+  test("Distributed PCA Estimation should match local one") {
+    sc = new SparkContext("local", "test")
+
+    val matRows = 1000
+    val matCols = 10
+    val dimRed = 5
+
+    // Generate a random Gaussian matrix.
+    val gau = new Gaussian(0.0, 1.0)
+    val randMatrix = new DenseMatrix(matRows, matCols, gau.sample(matRows*matCols).toArray)
+
+    // Parallelize and estimate the PCA.
+    val data = sc.parallelize(MatrixUtils.matrixToRowArray(randMatrix).map(x => convert(x, Float)))
+
+    val pcaDist = new DistributedPCAEstimator(dimRed).fit(data)
+    val pcaLocal = new PCAEstimator(dimRed).fit(data)
+
+    assert(Stats.aboutEq(convert(pcaDist.pcaMat, Double), convert(pcaLocal.pcaMat, Double), 1e-4))
+  }
+
+  test("Approximate PCA Estimation should match local one") {
+    sc = new SparkContext("local", "test")
+
+    val matRows = 10000
+    val matCols = 100
+    val dimRed = 5
+
+    // Generate a random Gaussian matrix.
+    val gau = new Gaussian(0.0, 1.0)
+    val randMatrix = new DenseMatrix(matRows, matCols, gau.sample(matRows*matCols).toArray)
+
+    // Parallelize and estimate the PCA.
+    val data = sc.parallelize(MatrixUtils.matrixToRowArray(randMatrix).map(x => convert(x, Float)))
+
+    val pcaApprox = new ApproximatePCAEstimator(dimRed, q = 500).fit(data)
```

Comment:
Basically the runtime scales with q(k+p)^2 - so a really big p is a bad
idea, but if it's small (say 5 or 10) it's probably ok.

On Fri, Dec 18, 2015 at 10:39 AM, Tomer Kaftan <notifications@github.com>
wrote:

> In src/test/scala/nodes/learning/PCASuite.scala
> <https://github.com/amplab/keystone/pull/194#discussion_r48054761>:
>
> > +
> > +  test("Approximate PCA Estimation should match local one") {
> > +    sc = new SparkContext("local", "test")
> > +
> > +    val matRows = 10000
> > +    val matCols = 100
> > +    val dimRed = 5
> > +
> > +    // Generate a random Gaussian matrix.
> > +    val gau = new Gaussian(0.0, 1.0)
> > +    val randMatrix = new DenseMatrix(matRows, matCols, gau.sample(matRows*matCols).toArray)
> > +
> > +    // Parallelize and estimate the PCA.
> > +    val data = sc.parallelize(MatrixUtils.matrixToRowArray(randMatrix).map(x => convert(x, Float)))
> > +
> > +    val pcaApprox = new ApproximatePCAEstimator(dimRed, q = 500).fit(data)
>
> Hmm, looks like if I increase p when I lower dimReds it'll also work
> better (but I have no sense of what implications that has on performance /
> etc.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/keystone/pull/194/files#r48054761>.
>